helix-rest/src/main/java/org/apache/helix/rest/metrics/JmxReporter.java

```diff
@@ -0,0 +1,815 @@
+package org.apache.helix.rest.metrics;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.Closeable;
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.Hashtable;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import javax.management.InstanceAlreadyExistsException;
+import javax.management.InstanceNotFoundException;
+import javax.management.JMException;
+import javax.management.MBeanRegistrationException;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectInstance;
+import javax.management.ObjectName;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Metered;
+import com.codahale.metrics.MetricFilter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.MetricRegistryListener;
+import com.codahale.metrics.Reporter;
+import com.codahale.metrics.Timer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A reporter which listens for new metrics and exposes them as namespaced MBeans.
+ * <p>
+ * The code is largely based on the {@code JmxReporter} class of the dropwizard metrics library
+ * <pre>See
+ * <a href="https://github.com/dropwizard/metrics/blob/master/metrics-core/src/main/java/io/
+ * dropwizard/metrics/JmxReporter.java">JmxReporter</a>
+ * </pre>
+ */
+public class JmxReporter implements Reporter, Closeable {
```

Comment:
How many codes do we need to change?
1. is it possible we extend the original JmxReporter?
2. is it possible you highlight the changed logic?
This file is not quite reviewable.

Reply:
1. It is not possible that we extend the original JmxReporter, because it has internal classes that are private.
2. I put the the change description in the PR description: 
Some changes are applied:
* Add withKeyProperties() to builder so we can add more key properties to an object name
* Remove the ObjectNameFactory and put the simple logic to the internal createName()
* Fix and add unit tests

But let me also add some comments to highlight our change.

Reply:
Based on what you described, I have a simpler alternative design that might work.

We implement our own ObjectNameFactory instead of JmxReporter. Let's call it NamespaceObjectNameFactory. It will be initialized with the namespace. Then you can put the namespace to the ObjectName whenever the reporter calls CreateName method.

Reply:
I understand if we only want to customize the object name, we don’t have to add Jmx reporter. I was thinking we firstly implement our object name factory. Of course, like you said, we can put the key properties in the constructor: `ObjectNameFactory(HashTable<String, String>())`. And later if we want to remove those attributes, we add jmx reporter.
It really depends on whether if we need to remove some attributes. We can do it now or later. I don’t have a strong preference, but if we need to remove the attributes, we need to have the JmxReporter.